From d6c01898e1de8a2abffb1acea60769ae077c7255 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:01:34 -0400
Subject: [usb] wusbcore: implement hwahc_op_get_frame_number

Message-id: <1410887463-20674-23-git-send-email-dzickus@redhat.com>
Patchwork-id: 95112
O-Subject: [RHEL7 PATCH 022/591] usb: wusbcore: implement hwahc_op_get_frame_number
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 7e176dcb035eef8917e5fc2e653460b5410dbc34
Author: Thomas Pugliese <thomas.pugliese@gmail.com>
Date:   Tue Oct 1 14:04:33 2013 -0500

    usb: wusbcore: implement hwahc_op_get_frame_number

    This patch adds an implementation for hwahc_op_get_frame_number.  The
    request is fulfulled by forwarding it to the lower hcd.  This was done
    because the GET_TIME request on the HWA requires sending an URB to the
    HWA and waiting for the results which cannot be done in atomic context.

    Signed-off-by: Thomas Pugliese <thomas.pugliese@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/hwa-hc.c b/drivers/usb/host/hwa-hc.c
index 5b86ffb..e58b924 100644
--- a/drivers/usb/host/hwa-hc.c
+++ b/drivers/usb/host/hwa-hc.c
@@ -199,10 +199,14 @@ static int hwahc_op_get_frame_number(struct usb_hcd *usb_hcd)
 {
  struct wusbhc *wusbhc = usb_hcd_to_wusbhc(usb_hcd);
  struct hwahc *hwahc = container_of(wusbhc, struct hwahc, wusbhc);
+ struct wahc *wa = &hwahc->wa;
 
- dev_err(wusbhc->dev, "%s (%p [%p]) UNIMPLEMENTED\n", __func__,
-  usb_hcd, hwahc);
- return -ENOSYS;
+ /*
+  * We cannot query the HWA for the WUSB time since that requires sending
+  * a synchronous URB and this function can be called in_interrupt.
+  * Instead, query the USB frame number for our parent and use that.
+  */
+ return usb_get_current_frame_number(wa->usb_dev);
 }
 
 static int hwahc_op_urb_enqueue(struct usb_hcd *usb_hcd, struct urb *urb,
-- 
1.7.1